pyfr/nputil.py

```diff
@@ -126,3 +127,187 @@ def npdtype_to_ctypestype(dtype):
         return None
 
     return _ctypestype_map[np.dtype(dtype).type]
+
+
+#Get indices for interval segments of interpolation
+def get_interval_idxs(t, xarr):
+    if np.asarray(xarr).size > 1:
+        idxs = np.array([np.max(np.nonzero(xx >= t[:-1] - 1e-12)[0]) 
+                                                        for xx in xarr])
+    else:
+        idxs = np.max(np.nonzero(xarr >= t[:-1] - 1e-12)[0])
+    return idxs
+
+#Modified Gauss Elimination/Thompson algorithm for tridiag systems
+#All changes are done in place and the algorithm 
+#is applied to the lhs and rhs together
+def trislv(a, b, c, d):
+    #forward elimination
+    d[0] /= b[0]
+    for i, (ai, bi, bj, ci, dj) in enumerate(
+                zip(a, b[1: ], b[: -1], c, d[: -1]), start=1):
+        #lhs coeffs
+        c[i-1] = cdivb = ci/bj
+        b[i] = bi1 = bi - ai*cdivb
+        #rhs     
+        d[i] -= ai*dj
+        d[i] /= bi1
+
+    #backward substitution
+    dd = d[::-1]
+    for i, (ci, dj) in enumerate(zip(c[::-1], d[:0:-1]), start=1):
+        dd[i] -= ci*dj
+
+
+#tridiagonal solver for cyclic/periodic systems 
+#Using Sherman-Morrison formula and Thompson algorithm
+# A = A + u ⊗ v
+# u, v are vectors
+def trislvcyc(a, b, c, d, alpha, beta):
+    u = np.zeros(d.shape[0], dtype=d.dtype)
+    #Solve the two systems, Ax = d, Bz = u in place, i.e., d->x, u->z
+    gam = -b[0] 
+    b[0] *= 2
+    b[-1] -= alpha*beta/gam
+    u[0] = gam
+    u[-1] = alpha
+    #solve the two systems in one loop
+    #forward elimination
+    d[0] /= b[0]
+    u[0] /= b[0]
+    for i, (ai, bi, bj, ci, dj, uj) in enumerate(
+                zip(a, b[1: ], b[: -1], c, d[: -1], u[: -1]), start=1):
+        #lhs coeffs
+        c[i-1] = cdivb = ci/bj
+        b[i]  = bi1 = bi - ai*cdivb  
+        # first rhs
+        d[i] -= ai*dj
+        d[i] /= bi1
+        # second rhs
+        u[i] -= ai*uj
+        u[i] /= bi1
+
+    #backward substitution
+    dd = d[::-1]
+    uu = u[::-1]
+    for i, (ci, dj, uj) in enumerate(
+            zip(c[::-1], d[:0:-1], u[:0:-1]), start=1):
+        dd[i] -= ci*dj
+        uu[i] -= ci*uj
+
+    #compute the final solution
+    fact = (d[0] + alpha*d[-1]/gam)/(1.0 + u[0] + alpha*u[-1]/gam)
+    d -= fact*u[:, None]
+
+    
+class CurveFit:
+    def __init__(self, *args, **kwargs):
+        pass
+
+    def intrp(self, *args):
+        pass
+
+class LinearFit(CurveFit):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self._interp = np.vectorize(np.interp, signature='(r),(n),(m)->(r)')
```

Comment:
Why is this needed?  Why not call the relevant function in a loop inside interp?

Reply:
both do the same thing, either a loop or using vectorize.